### DISCUSSION TITLE
Code consistency in tutorials, examples, etc.

### DISCUSSION THREAD

**Discussion by aeosynth** [ratatui/ratatui#614] [Discussion]
the same code is written differently in different places, and as a new user, i'd much rather have more consistency. just one example:

https://ratatui.rs/tutorial/counter-app/multiple-functions.html

```rust
fn main() -> Result<()> {
  startup()?;
  let status = run();
  shutdown()?;
  status?;
  Ok(())
}
```

```rust
fn main() -> Result<()> {
  startup()?;
  let result = run();
  shutdown()?;
  result?;
  Ok(())
}
```

**Comment by joshka** [comment]
I'm going to move this to the main repo and convert it to a discussion as I'd like to expand this with a more all encompassing answer.

**Comment by joshka** [comment]
In general we've got a lot of this across the various docs and examples. It happens because there's no obviously "correct" way to architect a console application, and so what works has evolved over time as various people have contributed fixes for the parts that frustrate them. This is the nature of a library (your code calls us) vs. framework (we call your code). We started with very little tutorial documentation earlier this year, and now we have a lot more quantity generated by a few different people. That's led to at least 3 ways of doing things (probably more even). So now's the time to consolidate the quantity into high quality consistent docs.

Generally, the following things are implemented subtly differently across applications:
- terminal setup / restore in the main function vs separate functions
- application main loop in the main function or in another function
- application type is called by the main loop or app type contains the main loop
- terminal usage as functions vs encapsulated in a wrapper type
- simple things like using stderr vs stdout
- using `stdout().execute()` vs `execute!()` for crossterm calls
- manual terminal restore vs automatic Drop implementation
- having tick as a concept vs rendering at a fixed FPS vs having a specific dirty flag or update event/action
- timeouts / ticks handled as part of the event handling or seperate
- async vs polling vs threads
- the above code where we're saving a result and later calling shutdown before unwrapping is not very idiomatic rust
- using style shorthands (`"foo".blue()`) vs long form (`Span::styled("foo", Style::default().fg(Color::Blue))`)
- etc.

I definitely agree that we should be more consistent and show a single approach in all the examples, tutorials, how-tos etc. So I'd like to suggest some general things that we can incorporate into our examples as standard. The overarching goal should be to show a single consistent way for all examples, and then only show a different way as a how-to specific to that other way.

1. Terminal setup / restore should be handled in a way that means that the terminal is never left in a broken state due to raw mode / alternate screen (even for the simplest application). Right now that requires creating a wrapper class that implements Drop, but https://github.com/ratatui-org/ratatui/pull/280 will make this a standard part of Ratatui (WIP).
2. If we have an App type, then the app type has a run method which is main driver (i.e. we don't see methods that take `&app` as a parameter, instead we see methods on `App` which have `&self` as a parameter. I really like the approach that the implementation of `fn main()` is just `App::run()?` (with perhaps some extra for logging init etc.) as this means that we have a terminal that restores properly and any results / panics are going to be printed to the console correctly.
3. We handle panic correctly in every app example
4. We pick and use a result lib instead of `Result<T,E>` (probably color-eyre is the best available for this)
5. We pick and use a command arg parser (clap is the most used)
6. Use stdout over stderr (due to rendering speed differences)

I'm sure there are other things to note that are inconsistent.

How does this sound?
